### ISSUE TITLE
The second-order gradient of tf.math.log and tf.math.log1p is 0 when the input is NaN.

### ISSUE THREAD

**Issue by maybeLee** [tensorflow/tensorflow#59046]
<details><summary>Click to expand!</summary> 
 
 ### Issue Type

Bug

### Source

binary

### Tensorflow Version

2.9.2

### Custom Code

Yes

### Python version

3.8

### Current Behaviour?

```shell
I was trying some loss functions and I find when the output of `tf.math.log1p` and `tf.math.log` is NaN, second-order gradients of one variable in these functions are still 0.0. Here is the equation of the loss function that causes this issue:

res = tf.math.log(tf.math.sqrt(1-x)-y+y)
```

When `1-x<0`, `tf.math.sqrt` outputs `NaN` and `res` is also `NaN`. However, the second-order derivative of `y` in this equation is `0.0`. Since `NaN` indicates some potential errors during model training and this special value should be manifested to the derivative, I think output `0.0` when receiving `NaN` is not optimal.

For your reference, I also tried on PyTorch and found that the second-order derivative of `y` in PyTorch is `NaN`, which I think would be better.

Please note that this issue happens in both `tf.math.log` and `tf.math.log1p` operators, while does not occur in operators such as `tf.math.sin`. Specifically, if I change the equation to `tf.math.sin(tf.math.sqrt(1-x)-y+y)`, the second-order derivative of y is `NaN`.
```


### Standalone code to reproduce the issue

```shell
Access this colab link (https://colab.research.google.com/drive/1UvQFl6wyvcrfqfu_x1lbOFVTHmO0n0bO?usp=sharing) or run below code to reproduce the issue. 


import tensorflow as tf
x = tf.Variable(100.)
y = tf.Variable(100.)
with tf.GradientTape() as g1:
  with tf.GradientTape() as g0:
    g0.watch(y)
    res = tf.math.log(tf.math.sqrt(1-x)-y+y)
  first_order_gradient = g0.gradient(res, y)
second_order_gradient = g1.gradient(first_order_gradient, y)
    
print("TensorFlow's output is: ", first_order_gradient, second_order_gradient)
```

```
import torch

x = torch.tensor(100., requires_grad=True)
y = torch.tensor(100., requires_grad=True)
res = torch.log(torch.sqrt(1-x)-y+y)
first_order_gradient = torch.autograd.grad(res, (y,), create_graph=True)
second_order_gradient = torch.autograd.grad(first_order_gradient, (y,), create_graph=True)

print("PyTorch's output is: ", first_order_gradient, second_order_gradient)
```
```


### Relevant log output

```shell
TensorFlow's output is:  tf.Tensor(nan, shape=(), dtype=float32) tf.Tensor(0.0, shape=(), dtype=float32)

PyTorch's output is:  (tensor(nan, grad_fn=<AddBackward0>),) (tensor(nan, grad_fn=<AddBackward0>),)
```
</details>

**Comment by tilakrayal** [comment]
@sushreebarsa,
I was able to reproduce the issue on tensorflow [v2.9](https://colab.research.google.com/gist/tilakrayal/fd585b144ac4bc2f5b5fb3f8a35a14c1/tf-log1p-log-derivatives-0-0.ipynb), [v2.11](https://colab.research.google.com/gist/tilakrayal/5b67f8c68d8f66b5bd6b9eec63a8a105/untitled824.ipynb) and [nightly](https://colab.research.google.com/gist/tilakrayal/56ad1f853f62e1c25b1812f9c39240d6/untitlenightly.ipynb). Kindly find the gist of it here.

**Comment by cantonios** [comment]
This is a bit of an oddball.

Your input function is _actually_ independent of `y`, since analytically
```
log(sqrt(1-x)-y+y)) = log(sqrt(1-x))
```
So technically, the first derivative w.r.t. y should be zero.  Except that we don't do analytical math, so we end up with a `NaN * 0 = NaN`.

For the second derivative, we have special handling inside the reciprocal gradient (reciprocal being the gradient of log) that avoids generating NaNs when passing zero gradients through the singularity at 0.  This is where we are returning 0, and where something like `sin` does not.  I _can_ make that consistent to return NaN.  But really, your gradients should technically be zero.
